pkg/api/controller.go

```diff
@@ -2334,85 +2338,156 @@ func (c *Controller) StageObject(w http.ResponseWriter, r *http.Request, body St
+	if srcRef == branch {
+		copyType = httpHeaderCopyTypeShallow
+		// try getting source entry from staging - if not found we continue to fall to full copy
+		srcEntry, err = c.Catalog.GetEntry(ctx, repository, branch, srcPath, catalog.GetEntryParams{
+			ReturnExpired: true,
+			StageOnly:     true,
+		})
+		if !errors.Is(err, graveler.ErrNotFound) && c.handleAPIError(ctx, w, r, err) {
+			return
+		}
+		if srcEntry != nil {
+			entry, err = c.copyObjectShallow(ctx, repository, branch, srcEntry, destPath)
+			// for ErrTooManyTries we like to continue falling to full copy
+			if !errors.Is(err, graveler.ErrTooManyTries) && c.handleAPIError(ctx, w, r, err) {
+				return
+			}
+		}
 	}
-	qk, err := block.ResolveNamespace(repo.StorageNamespace, blob.PhysicalAddress, block.IdentifierTypeRelative)
+
+	// full copy if no entry was created until here
+	if entry == nil {
+		copyType = httpHeaderCopyTypeFull
+		entry, err = c.copyObjectFull(ctx, repo, branch, srcPath, destPath, srcEntry)
+		if c.handleAPIError(ctx, w, r, err) {
+			return
+		}
+	}
+
+	qk, err := block.ResolveNamespace(repo.StorageNamespace, entry.PhysicalAddress, block.IdentifierTypeRelative)
 	if err != nil {
 		writeError(w, r, http.StatusInternalServerError, err)
 		return
 	}
-	w.Header().Set("X-lakeFS-Copy-Type", "full") // update with the real implementation
-	// TODO: end of TODO
 
+	w.Header().Set(httpHeaderCopyType, copyType)
 	response := ObjectStats{
 		Checksum:        entry.Checksum,
 		Mtime:           entry.CreationDate.Unix(),
 		Path:            entry.Path,
 		PathType:        entryTypeObject,
 		PhysicalAddress: qk.Format(),
 		SizeBytes:       Int64Ptr(entry.Size),
-		ContentType:     &entry.ContentType,
+		ContentType:     StringPtr(entry.ContentType),
 	}
 	writeResponse(w, r, http.StatusCreated, response)
 }
 
+func (c *Controller) copyObjectShallow(ctx context.Context, repository, branch string, srcEntry *catalog.DBEntry, destPath string) (*catalog.DBEntry, error) {
+	// track physical address (copy-table)
+	_, err := c.Catalog.TrackPhysicalAddress(ctx, repository, srcEntry.PhysicalAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	// create entry (copy) - try only once
+	dstEntry := *srcEntry
+	dstEntry.CreationDate = time.Now().UTC()
+	dstEntry.Path = destPath
+	err = c.Catalog.CreateEntry(ctx, repository, branch, dstEntry, graveler.WithMaxTries(1))
+	if err != nil {
+		return nil, err
+	}
+	return &dstEntry, nil
+}
+
+// copyObjectFull copy data from srcEntry's physical address (if set) or srcPath into destPath
+func (c *Controller) copyObjectFull(ctx context.Context, repo *catalog.Repository, branch string, srcPath string, destPath string, srcEntry *catalog.DBEntry) (*catalog.DBEntry, error) {
+	var err error
+	// fetch src entry if needed - optimization in case we already have the entry
+	if srcEntry == nil {
+		srcEntry, err = c.Catalog.GetEntry(ctx, repo.Name, branch, srcPath, catalog.GetEntryParams{ReturnExpired: true})
```

Comment:
Why do we return expired?

Reply:
This is the default for get object - but looking into it, no one is using this field so I'll handle it in a different PR.

Reply:
Handled in this PR